### ISSUE TITLE
rkt networking discussion

### ISSUE THREAD

**Issue by euank** [kubernetes/kubernetes#26102]
There are a number of small pieces of discussion scattered across this and the rkt repository related to how to handle networking setup; I'd like to consolidate this discussion here, talk about all the issues and options I know about, and hopefully figure out what the best option is.

Discussion refs:
https://github.com/coreos/rkt/issues/2249
https://github.com/kubernetes/kubernetes/issues/26081
https://github.com/kubernetes/kubernetes/issues/25404#issuecomment-218402169
https://github.com/kubernetes/kubernetes/pull/24688
https://github.com/kubernetes/kubernetes/issues/21923
And quite possibly several more I missed
## Introduction

Kubernetes and rkt both make use of CNI to configure container network namespaces. Each has independent means of configuring and executing plugins. Kubernetes also has a number of features that relate somewhat to the network namespace, such as availability of the pod's IP in a number of locations (status, downwards API, etc).

Until recently, rkt's handled network configuration on its own with no regard for Kubernetes' network logic. As of #25062, it was changed such that kubelet handles network setup and configuration for rkt. Unfortunately, both of these solutions have problems.

Below, the problems with each approach are listed. Finally, some paths forwards are suggested.
## Issues
- The PodIP should be available to the kubelet prior to applications running for a number of features to work
  - Downwards API (both volume and environment variables) rely on this
  - `/etc/hosts` file, created by kubelet, relies on this (alternately, rkt should create this correctly)
- The network plugin configuration used by the docker container runtime should work, in general, for rkt with no changes (including filepath) (https://github.com/coreos/rkt/issues/2249)
- The solution should work with any rkt stage1 (https://github.com/kubernetes/kubernetes/issues/26081)
## Solutions
1. Rkt manages the network
2. Kubelet manages the network
3. Hybrid of 1 & 2
### Rkt manages networking

This option has a few problems:
1. The PodIP is not available to the kubelet until after the application is finalized (including mounts and environment variables) (https://github.com/coreos/rkt/issues/2223 and https://github.com/coreos/rkt/issues/2375)
2. The network plugin configuration path is not configurable (https://github.com/coreos/rkt/pull/2312)
3. Duplication of logic/code between kubelet and rkt (more or less depending on who creates `/etc/hosts` and so on).

The first issue there is a difficult one to solve and is the main reason that this approach was discarded, whether temporarily or not.

This is probably the only robust long-term solution when considering the large variety of stage1's (and thus network configurations) that are available and, once rkt provides a container-level interface (see https://github.com/coreos/rkt/issues/2375), there should be a natural way to solve problem number one via the separation of sandbox creation and container creation.
### Kubelet manages networking

This option has its own problems:
1. Does not work with kvm flavor stage1 (or any stage1 that does not use Linux network namespaces or does not support `--net=host`).
2. Depends on availability of nsenter / ip tools (already depended on elsewhere I believe, mostly a nonissue)
3. Results in the `rkt` cli reporting incorrect network information

This allows fulfilling more of kubelet's assumptions about networking, but does not play nicely with rkt's existing abstractions. It does solve the most problems for the case of the default stage1 and existing rkt featureset while also not being a large amount of code. It also shares significant logic with the docker network setup, with the benefits that entails.
### Hybrid

The kubelet could decide whether it or rkt should manage the networking based on whether the stage1 suports it. When rkt manages networking, the features dependent on PodIP being available early on will not work.
1. More code in kubelet to support both methods
2. rkt does not provide an easy means to determine whether one or the other is appropriate currently
3. Inconsistent featureset depending on stage1 (:frowning:)
## Conclusion

My current preference is to continue with the kubelet-managed network for 1.0 (though given the timeline, I'm not sure there's another choice), and rethink this afterwards. If it's possible to support a hybrid approach by 1.0 with the understanding that it's a temporary solution, that also seems reasonable.

Ideally, the kubelet's abstractions would not interfere with rkt's ability to manage networking itself since, I admit, the above solution is a bit of a hacky one.

It's also possible that the upcoming Container Runtime Interface changes, with a focus on sandboxes, will require further changes in both the kubelet and rkt that changes what issues are present.

I'd like to gather other's thoughts and make sure we're all on the same page and figure out if there are other options or issues I'm not thinking of.

/cc @yifan-gu @jellonek @pskrzyns @jonboulle @dcbw @iaguis @philips @thockin @freehan

**Comment by jellonek** [comment]
Very good and detailed summary.

Because of mentioned narrowing of `rkt` possibilities I have to opt on first proposal.

Second option looks a bit like usage of ORM which have abstraction for multiple databases, but is using only features found in each of them, so e.x. not using foreign keys given by postgres, because we have similar functionality on ORM side prepared for databases like mysql, which do not support foreign keys in reasonable way.

**Comment by dcbw** [comment]
First, I (and others) are aware that the current Runtime object model doesn't work well for VM-style runtimes and I know ther'es ongoing work to fix that.  I don't think that precludes using Kubernetes network plugins for VMs as long as the net plugin API is sufficient generalized.  The only pain point I can think of right now is GetNetNS() which assumes a network namespace.  That's clearly not going to be the case for VM-style runtimes, but right now CNI currently assumes network namespace path so we'll need to change the CNI spec to make that more generic, in addition to changing the Runtime interface to not assume pods are running in a VM.

Second I believe there's value in having network setup entirely separate from the runtime. No matter which of docker, rkt, or runc are used as the pod engine, I may wish to use a 3rd party network plugin, and that plugin's authors probably don't want to write a couple different shims for each container runtime just so it works with Kubernetes.  I guess I'm not sure why, from a user perspective, kubernetes-on-rkt should work any differently from kubernetes-on-docker from kubernetes-on-runc...

On the specific downsides listed for Kubernetes managed networking:

> Does not work with kvm flavor stage1 (or any stage1 that does not use Linux network namespaces or does not support --net=host).

see above...

> Depends on availability of nsenter / ip tools (already depended on elsewhere I believe, mostly a nonissue)

Yeah, already depended on elsewhere, but something I don't really like.  Now that CNI's pkg/ns is mostly fixed up I'd really like to use that instead of execers.

> Results in the rkt cli reporting incorrect network information

I'd argue this is a larger problem with container runtimes; the same problem exists with Docker.  One could just as easily argue that when running pods under orchestration systems like Kubernetes, anything that wants to know the container/pod's IP address should be asking the orchestration layer instead of the container runtime, because the container runtime does not always manage the network configuration.  If the pod is orchestrated, the container runtime clearly isn't going to have all the information, but if people really insist on having the container runtime also present networking details, perhaps there needs to be a way for higher layers to send that information down to the runtime itself so 'docker inspect' and 'rkt status' work as before.

...

That all said, I personally don't care what manages networking as long as (a) kubelet itself doesn't do it (which is my end goal, of which kubenet and CNI are steps toward that goal), (b) plugin authors don't have to rewrite for every container runtime, and (c) there's consistent operation for cluster administrators and application authors no matter what runtime they choose to use.

**Comment by jellonek** [comment]
FYI: kvm flavored rkt does not use all CNI network plugins, only IPAM plugins. Rest of them (flannel, bridge, macvlan...) are reimplemented in `rkt` code to function like they. Because of that, current implementation requires that network should be configured internally by `rkt`.
We could discuss about additional step, like already existing "prepare", but for networking. This would help a bit with mentioned problem with `PodIP`.

Host networking is not (at least now) possible with 2 separate kernels. So there is no way to have network setup separated from this runtime.

**Comment by steveeJ** [comment]
Thanks for summarizing this @euank. 

@dcbw 
> right now CNI currently assumes network namespace path so we'll need to change the CNI spec to make that more generic

This is merely an implementation restriction and not enforced by the SPEC. In https://github.com/containernetworking/cni/pull/230 that restriction already got loosened a bit, even though there is more work needed. E.g., plugins should be able to declare support/requirements for such things.

>> Depends on availability of nsenter / ip tools (already depended on elsewhere I believe, mostly a nonissue)
> 
> Yeah, already depended on elsewhere, but something I don't really like. Now that CNI's pkg/ns is mostly fixed up I'd really like to use that instead of execers.

+1 on not depending on those tools, however executing external binaries is not wrong in general, after all CNI plugins are designed to be called like that.

> One could just as easily argue that when running pods under orchestration systems like Kubernetes, anything that wants to know the container/pod's IP address should be asking the orchestration layer

I remember having a similar discussion at the end of 2015 within a CNI/rkt sync, but the result was not conclusive and this still seems to case from k8s perspective. In order to solve this, everybody needs to agree on a common format and place to collect network metrics about pods/containers. If the runtimes and the orchestrater are all compatible with that, the runtime can be indicated to run in a let's call it _managed networking_ mode and extract networking information for various reasons, e.g. configuring DNS, hosts file, etc. I will repeat my key point here, we need to decide about the goal and then establish a solid communication to achieve compatibility.

@jellonek 
> kvm flavored rkt does not use all CNI network plugins, only IPAM plugins. Rest of them (flannel, bridge, macvlan...) are reimplemented in rkt code to function like they. Because of that, current implementation requires that network should be configured internally by rkt.

This is another implementation restriction. I can't think of a technical reason why the tap devices can't be setup by an orchestrator or in pure runtime case be delegated to an external component similar to how it's done with rkt nspawn+CNI. Of course this would need new or modified plugins but it would help to solve the problem we're facing now. Also, it could reduce duplication in case we'll see more runtimes that are based on hypervisors which use tap devices for network functionality.


TL;DR
I'm in favor of a hybrid model, in a way that the runtime knows it is in a managed network mode and can access, understand and use the network information, which must be stored in a copmatible format.

**Comment by jellonek** [comment]
@steveeJ I'm also sure that there is no technical reason why orchestrator could not prepare tap devices - I described only a current implementation. Only limitation which could be seen now - there is no possibility to mimic host networking using 2 kernels.

**Comment by feiskyer** [comment]
@jellonek @steveeJ @dcbw  I'm working on integrating `HyperContainer` (a hypervisor-based container runtime) into kubernetes. Tap device is indeed a pre-requirement for this runtime.

I'd like to add another network plugin to support hypervisor-based container runtimes, which will create a tap-device and return the network information (same as https://github.com/containernetworking/cni/pull/145) to runtime. This will also benefits kvm flavored rkt.

@dcbw I think current network plugin may also needs information as specified in https://github.com/containernetworking/cni/pull/145.

**Comment by jellonek** [comment]
As I announced on IRC channel - I'll try to prepare PR with proposal of interface for cgroups/vm based containerization engines, for first of them returning actual type of result, for second one - configuring tap devices and appending their device file locations into rest of result dictionary.
(with initial reference to https://github.com/containernetworking/cni/blob/master/SPEC.md - in mentioned PR there will be good place to discuss detail about particular key/value correct place).

Information about which kind of configuration should be prepared is planned as additional environment value with name prefixed by `CNI_`, so this would be independent from  "network configuration" json which still could be reused between different containerization engines.

Timeline for this proposal is quite long, because of my current job transition (so this should be done in time up to ~4 weeks).

**Comment by feiskyer** [comment]
@jellonek Great work. Ping me pls when the PR is ready.

**Comment by tomdee** [comment]
#containernetworking on freenode - see https://github.com/containernetworking/cni#contact

**Comment by pskrzyns** [comment]
cc @jjlakis @lukasredynk

**Comment by euank** [comment]
xref #27015 .. That pull request modifies the default behaviour when no network plugin is specified to use whatever rkt's configured with.

It's solution 3 above. As stated above, this isn't a perfect solution (downwards api / etc/hosts when using rkt's networking, need to configure network plugins out of band of kubelet when using rkt's networking).
This discussion has been moved to rktnetes 1.1 in the hopes we'll end up with a better solution.

**Comment by pskrzyns** [comment]
I think we should use 1st solution - Rkt manages the network
so this would be looking like that:
![cni](https://cloud.githubusercontent.com/assets/16152568/16118579/0a1619be-33d8-11e6-8754-51195b310149.png)
This is quite generic so other container solutions can use similiar approach

**Comment by feiskyer** [comment]
@jellonek The proposal to support hypervisor runtime in CNI:  https://github.com/containernetworking/cni/issues/251

**Comment by dcbw** [comment]
> I think we should use 1st solution - Rkt manages the network
> so this would be looking like that:
> cni
> This is quite generic so other container solutions can use similiar approach

As I mentioned above, the downside here is that the network plugin is tied to the runtime.  You cannot use the same network plugin with whatever runtime you choose, let's say docker or rkt.  You must either specify somehow out-of-band via rkt- or docker- specific mechanisms what network plugin this particular pod needs, and/or the network plugin itself must have specific code for either docker or rkt.  That's not great for either cluster administrators or plugin authors.

Yes, this position does somewhat ignore the VM discussion, since obviously the existing k8s kubenet and CNI plugins cannot configure a VM.  But that doesn't mean we can't make them do that, eventually.

**Comment by pskrzyns** [comment]
@dcbw we are going to extend CNI to support VM configuration
https://github.com/containernetworking/cni/issues/251#issuecomment-231378414
https://github.com/coreos/rkt/issues/2791#issuecomment-231348752

I dont get it why i cannot use same plugin for rkt and docker for both of them (without hypevisor) plugin just need to prepare namespace and runtime need to know how to use it or i'm missing something ?

**Comment by steveeJ** [comment]
@pskrzyns 

> I dont get it why i cannot use same plugin for rkt and docker for both of them (without hypevisor) plugin just need to prepare namespace and runtime need to know how to use it or i'm missing something ?

@dcbw was responding to your previous comment that suggested to use solution 1 (rkt and respectively Docker manages the network). In that case, it wouldn't be the same plugin but independent implementations of the network plugins in order to accomplish the same thing. Kubernetes would need to pass runtime-specific network options to each runtime and compatibility would be very dependent on the runtime's implementation.

**Comment by pskrzyns** [comment]
Basing on https://github.com/kubernetes/kubernetes/issues/28667 especially https://github.com/kubernetes/kubernetes/issues/28667#issuecomment-238771135 we will stick with 1 solution (Rkt manages the network)
